### PR TITLE
feat: Auto performance tracing with XHR/fetch, and routing instrumentation

### PR DESCRIPTION
**Pull request by jennmueng**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

Note: Needs https://github.com/getsentry/sentry-javascript/pull/3144 to be released over in js before this can be merged.

See `App.tsx` in the sample as an example of how to initialize auto performance tracing.

Screen Shot 2021-01-10 at 9 49 33 PM

### Routing

The routing Instrumentation has to be different than how it's currently implemented for browser given how routing in React Native is handled different and is not global. We would need to create an instance where a route event can be emitted by a navigation router and be listened to by our integration.

#### Base/Manual Routing Instrumentation

_"Manual-Auto"_

Create an instance of `RoutingInstrumentation`, pass it to the tracing integration, and then call `onRouteWillChange` every time the route changes with the desired transaction context. This sets an idle transaction on the scope.

```typescript
// Construct a new instrumentation instance. This is needed to communicate between the integration and React
const routingInstrumentation = new Sentry.RoutingInstrumentation();

Sentry.init({
  ...
  integrations: [
    new Sentry.ReactNativeTracing({
      // Pass instrumentation to be used as `routingInstrumentation`
      routingInstrumentation,
      ...
    }),
  ],
})

const App = () => {
  // Call this before the route changes
  routingInstrumentation.onRouteWillChange(transactionContext)
};
```

#### React Navigation V5 Instrumentation

Below is an example of how to use the current implementation:
```typescript
// Construct a new instrumentation instance. This is needed to communicate between the integration and React
const reactNavigationV5Instrumentation = new Sentry.ReactNavigationV5Instrumentation();

Sentry.init({
  ...
  integrations: [
    new Sentry.ReactNativeTracing({
      // Pass instrumentation to be used as `routingInstrumentation`
      routingInstrumentation: reactNavigationV5Instrumentation,
      ...
    }),
  ],
})

const App = () => {
  // Create a ref for the navigation container
  const navigation = React.createRef<NavigationContainerRef>();

  // Register the navigation container with the instrumentation
  React.useEffect(() => {
    reactNavigationV5Instrumentation.registerNavigationContainer(navigation);
  }, []);

  return (
    // Connect the ref to the navigation container
    <NavigationContainer ref={navigation}>
      ...
    </NavigationContainer>
  );
};
```

Preferably, all the react component part can just be done in a custom hook

#### React Navigation V4 Instrumentation

Works almost exactly like V5 on the surface, although entirely different underneath.

```typescript
// Construct a new instrumentation instance. This is needed to communicate between the integration and React
const reactNavigationV4Instrumentation = new Sentry.ReactNavigationV4Instrumentation();

Sentry.init({
  ...
  integrations: [
    new Sentry.ReactNativeTracing({
      // Pass instrumentation to be used as `routingInstrumentation`
      routingInstrumentation: reactNavigationV4Instrumentation,
      ...
    }),
  ],
})

const App = () => {
  // Create a ref for the navigation container
  const appContainer = React.createRef();

  // Register the navigation container with the instrumentation
  React.useEffect(() => {
    reactNavigationV4Instrumentation.registerAppContainer(appContainer);
  }, []);

  return (
    // Connect the ref to the navigation container
    <AppContainer ref={appContainer} />
  );
};
```

### Idle Transaction
The idle transaction is carried over from the browser implementation of tracing. It needs to be exposed upstream from `@sentry/tracing`. An idle transaction should be started and set on the scope every time a route changes using a routing instrumentation.

### Fetch/XHR
Fetch/XHR instrumentation is carried over from the browser implementation, and it also needs to be exposed.

### React Profiler
Works out of the box.

https://sentry.io/organizations/sentry-sdks/discover/sentry-react-native:157803b6f5b94f91b5da33014911f181/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&query=&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1

### Need to expose from @sentry/tracing

Some things need to be exposed upstream from `@sentry/tracing`:
```
IdleTransaction,
startIdleTransaction,
registerRequestInstrumentation,
RequestInstrumentationOptions,
defaultRequestInstrumentationOptions,
```

## Remaining things to do/investigate
- [x] Manual Routing Instrumentation
- [x] React Navigation V5
- [x] React Navigation V4
- [x] Fetch/XHR request instrumentation
- [x] Test Everything on iOS Sim
- [x] Test Everything on Android Sim
- [x] Clean up code and comment

## :green_heart: How did you test it?

Tested locally on iOS simulator with new sample app.

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: src/js/tracing/reactnavigation.ts
Comment (giautm) [inline src/js/tracing/reactnavigation.ts:53]: Should add `PUSH` & `REPLACE` actions?

Comment (jennmueng) [reply src/js/tracing/reactnavigation.ts:53]: Yes we should account for those too

**Comment by giautm** [comment]
Hi @jennmueng, do you need any help to finish this PR? I'm so awaiting it's go live

**Comment by jennmueng** [comment]
> Hi @jennmueng, do you need any help to finish this PR? I'm so awaiting it's go live

Nothing that would need help, just need time to finish everything and test everything

**Comment by jennmueng** [comment]
Updated the v5 instrumentation to handle cases where a router with action types other than the default stack ones is used. It now creates an idle transaction on **every** successful navigation action dispatch, but it does not include route context just yet. It will wait for the navigation state to change, and then update the transaction with the route context. If the state change doesn't happen within 200ms, the transaction is then discarded.

It provides the option of setting a `shouldAttachTransaction` method that would be called after each state change to determine whether to discard this transaction or keep it. By default it sets a transaction on every new route that isn't a repeat of the last route.

**Comment by jennmueng** [comment]
Update: Working perfectly and ready for iOS, however I've run into an issue on Android where the current SDK (`3.2.0`) does not fully support transaction types. Currently going to try a workaround.

**Comment by jennmueng** [comment]
Temporary Hack to use the js `FetchTransport` to send transactions on Android for now. This will not be GA and instead just for an alpha. We will need to wait for transactions on FileObserver envelope support on Android

**Comment by jennmueng** [comment]
Now with instrumentation for React-Navigation V4. It one works differently than the V5 one on the inside, although extremely similar on the outside. This time we patch the router on the `AppContainer`.

**Comment by jennmueng** [comment]
Blocked by https://github.com/getsentry/sentry-javascript/pull/3144 over in JS as everyone there is out of office.

**Comment by bruno-garcia** [comment]
Maybe worth resolving conflict with `master` so it's ready to merge

**Comment by jennmueng** [comment]
> Maybe worth resolving conflict with `master` so it's ready to merge

Sure, but it would also need a bump to the @sentry/javascript dependencies too before we can merge it btw.

**Comment by jennmueng** [comment]
Just a note: The CI is failing due to things from @sentry/tracing not being exported yet. Need to wait for a new release over at JS before these will pass.

**Comment by janicduplessis** [comment]
Probably out of scope for this PR, but are there plans to look into also instrumenting the native side? I'm pretty sure RN has hooks for this. This way it would be possible to visualize the full rendering, including react js render, RN bridging, layout and native render.

**Comment by bruno-garcia** [comment]
> Probably out of scope for this PR, but are there plans to look into also instrumenting the native side? I'm pretty sure RN has hooks for this. This way it would be possible to visualize the full rendering, including react js render, RN bridging, layout and native render.

I dunno if @jennmueng already looked at it, but could you point us to the hooks we can use for that?

We're adding performance API to the native layers next, but we don't have a story for propagating transaction through the bridge yet.

I'm hoping that at least a span for the bridge call will be possible even without any support from the native SDKs.

**Comment by jennmueng** [comment]
@janicduplessis Yes we're definitely looking into instrumenting React Native way more in the near future. We would really appreciate if you know any of these hooks that you mentioned to see if we can instrument something there.

**Comment by janicduplessis** [comment]
On Android there's this gist that shows how to use the ReactMarker API https://gist.github.com/axemclion/f01cbb32fcb5b14bf6b4fd4594192825#file-perflogger-java, which should have some interesting info. Not sure about what is available on iOS. Are you on RN contributor discord? There are probably people there that can help too.

**Comment by bruno-garcia** [comment]
@janicduplessis Thanks for the link! Not sure about @jennmueng, I'm not on that discord but I'm on the `#react-native` channel of [Sentry's Discord server](http://discord.gg/sentry), would you like to join and we discuss this there?

**Comment by jennmueng** [comment]
Note: Can use `tracesSampler` instead of the `shouldSendTransaction`

File: src/js/transports/native.ts
Comment (marandaneto) [inline src/js/transports/native.ts:14]: you can now experiment with https://github.com/getsentry/sentry-java/releases/tag/4.0.0-alpha.3
that should resolve the problem and the hack could be removed

Comment (jennmueng) [reply src/js/transports/native.ts:14]: Ok thanks, I'll give it a try and I think we can have an alpha ready by like Monday.

File: src/js/tracing/utils.ts
Comment (HazAT) [inline src/js/tracing/utils.ts:9]: We don't use this function, or do we?

File: src/js/tracing/reactnativetracing.ts
Comment (HazAT) [inline src/js/tracing/reactnativetracing.ts:201]: Here when calling `secToMs` inline, you can just do it in the `adjustTransactionDuration` internally. So you don't need to import `secToMs` here.

**Review by HazAT — APPROVED** [review]
Few nits

File: src/js/tracing/utils.ts
Comment (jennmueng) [reply src/js/tracing/utils.ts]: No we don't, it was copied over from js at first, let me remove it.

**Pull request merged**
